docs/history-service.md

```diff
@@ -0,0 +1,274 @@
+- Submitting tasks for execution to the task execution framework.
+- Periodically persisting queue "ack levels" (this periodic checkpointing represents a compromise between the desire to minimize the number of already-processed tasks processed on shard reload, and the desire to minimize write load).
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+- When the History service starts on a host, [`Start()`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/history_engine.go#L288) is called for each shard owned by that host, which starts the per-shard queue processor for each of the queue types.
+- Each queue processor [starts](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/queues/queue_base.go#L281) multiple [`Readers`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/queues/reader.go#L51) which, in response to messages on a certain channel, submit tasks to be processed.
+- The per-shards QueueProcessor goroutines submit tasks for execution to the task execution framework implemented in [`common/tasks/`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/common/tasks).
+- The task-execution framework itself uses a goroutine pool that is shared across shards. Ultimately, it [calls](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/common/tasks/fifo_scheduler.go#L219) `Execute` and then either `Ack` or `Nack` on each task.
+
+- Each queue processor also starts a goroutine running `processEventLoop`.
+  There are two variants of `processEventLoop`: [one for immediate queues](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/queues/queue_immediate.go#L150) such as the Transfer queue, and [one for scheduled queues](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/queues/queue_scheduled.go#L199) such as the Timer queue. For both queue types, [`checkpoint()`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/queues/queue_base.go#L349) is called periodically, which persists queue ack levels.
+
+- [`executions` table schema in Cassandra](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/schema/cassandra/temporal/schema.cql#L7)
+- [handling](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/timer_queue_active_task_executor.go#L114) of the various types of timers in the Timer Task queue.
+
+   </details>
+   <br>
+
+#### Transfer Task queue
+
+A Transfer Task is available to be executed immediately.
+When a Transfer Task is executed, an RPC is made to the Matching Service to enqueue the Workflow Task or Activity Task specified by the Transfer Task.
```

Comment:
those are only two types of transfer task. it looks like there are six more (close, cancel, signal, start child, auto-reset, delete).

Reply:
Thanks for pointing this out @dnr and @alexshtin. I think I had not picked up on that. I've added
> (In addition to enqueuing Workflow and Activity tasks, there are several other types of Transfer Task.)